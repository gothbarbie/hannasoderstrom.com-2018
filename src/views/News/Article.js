// @flow
import React, { Component } from 'react'
import { Link, withRouter } from 'react-router-dom'
import { connect } from 'react-redux'

import styled from 'styled-components'

import { fetchRecently } from '../../actions'

import Wrapper from '../../components/Layout/Wrapper'
import LoadingSpinner from '../../components/LoadingSpinner'
import H1 from '../../components/Typography/H1'
import Small from '../../components/Typography/Small'

import type { RouterHistory } from 'react-router-dom'

const WrapperSmall = styled.section`
  max-width: 600px;
  margin: 5% auto;
  padding: 0 2rem;

  p {
    line-height: 150%;
    font-weight: 300;
    color: #4f4f4f;
    font-size: 18px;
    margin-bottom: 1rem;
  }

  img {
    width: 100%;
    height: auto;
  }
`

const Header = styled.header`
  margin-bottom: 2rem;
`

type Props = {
  loading: boolean,
  fetchRecently: Function,
  history: RouterHistory,
  recentlies: Array<{
    content: {
      rendered: string,
    },
    excerpt: {
      rendered: string,
    },
    date: string,
    title: {
      rendered: string,
    },
    media: {
      guid?: {
        rendered: string,
      },
    },
  }>,
}

export class Article extends Component<Props> {
  componentDidMount () {
    const url = decodeURI(window.location.pathname).split('/')
    const slug = url[2]
    const regex = new RegExp(/([A-Za-z0-9-])+/g)

    if (regex.test(decodeURI(slug))) {
      this.props.fetchRecently(slug)
    }
  }

  render () {
    if (this.props.loading) return <LoadingSpinner />

    if (!this.props.recentlies.length) return null

    const { content, date, title } = this.props.recentlies[0]

    const dateFormatted = new Date(date)
    const options = {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
    }

    return (
      <Wrapper>
        <Link data-test="back-link" to="/news">
          ‚Üê Back
        </Link>
        <WrapperSmall>
          <Header>
            <H1
              dangerouslySetInnerHTML={{ __html: title.rendered }}
              data-test="article-title"
            />
            <Small
              dangerouslySetInnerHTML={{
                __html: dateFormatted.toLocaleDateString('en-EN', options),
              }}
            />
          </Header>
          <main
            className="main"
            dangerouslySetInnerHTML={{ __html: content.rendered }}
          />
        </WrapperSmall>
      </Wrapper>
    )
  }
}

const mapStateToProps = ({ recentlies, loading }) => {
  return {
    recentlies,
    loading,
  }
}

const mapDispatchToProps = {
  fetchRecently,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Article))
