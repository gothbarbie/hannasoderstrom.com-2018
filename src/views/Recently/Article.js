// @flow
import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'

import styled from 'styled-components'

import { fetchRecently } from '../../actions'

import Wrapper from '../../components/Layout/Wrapper'
import Small from '../../components/Small'
import LoadingSpinner from '../../components/LoadingSpinner'

const WrapperSmall = styled.section`
  max-width: 600px;
  margin: 5% auto;
  padding: 0 2rem;
`

const H1 = styled.h1`
  font-size: 2rem;
  font-weight: bold;
`

type Props = {
  loading: boolean,
  fetchRecently: Function,
  recentlies: Array<{}>,
}

class Article extends Component<Props> {
  componentDidMount () {
    const url = decodeURI(window.location.pathname).split('/')
    const slug = url[2]
    const regex = new RegExp(/([A-Za-z0-9\-])+/g)

    if (regex.test(decodeURI(slug))) {
      this.props.fetchRecently(slug)
    }
  }

  render () {
    if (this.props.loading) return <LoadingSpinner />

    if (!this.props.recentlies.length) return <p>Not found.</p>

    const { content, title } = this.props.recentlies[0]

    return (
      <Wrapper>
        <Link to="/recently">‚Üê Back</Link>
          <WrapperSmall>
            <H1 dangerouslySetInnerHTML={{ __html: title.rendered }} />
              <div dangerouslySetInnerHTML={{ __html: content.rendered }} />
          </WrapperSmall>
      </Wrapper>
    )
  }
}

const mapStateToProps = ({ recentlies, loading }) => {
  return {
    recentlies,
    loading,
  }
}

const mapDispatchToProps = {
  fetchRecently,
}

export default connect(mapStateToProps, mapDispatchToProps)(Article)
