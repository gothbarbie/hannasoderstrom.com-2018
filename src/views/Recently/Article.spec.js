import React from 'react'
import { shallow } from 'enzyme'

import { Article } from './Article'

describe('views/Article', () => {
  let component
  let store
  let fetchRecently

  beforeEach(() => {
    fetchRecently = jest.fn()
    store = {
      dispatch: jest.fn(),
      getState: jest.fn(),
      subscribe: jest.fn(),
    }
    component = shallow(
      <Article fetchRecently={fetchRecently} recentlies={[]} store={store} />
    )
  })

  it('renders when no content', () => {
    expect(component).toMatchSnapshot()
  })

  it('renders with content', () => {
    component.setProps({
      recentlies: [
        {
          id: 1,
          date: '2016-04-13T12:17:09',
          title: { rendered: 'Hello' },
          content: { rendered: '<p>Hello world</p>' },
        },
      ],
    })
    expect(component).toMatchSnapshot()
  })

  describe('#fetchRecently', () => {
    it('has been called on mount', () => {
      expect(fetchRecently).toHaveBeenCalled()
    })
  })
})
