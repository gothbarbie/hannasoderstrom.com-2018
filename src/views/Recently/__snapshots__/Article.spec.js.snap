// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/Article renders when no content 1`] = `""`;

exports[`views/Article renders with content 1`] = `
<styled.div>
  <Link
    replace={false}
    to="/recently"
  >
    ‚Üê Back
  </Link>
  <styled.section>
    <header>
      <styled.h1
        dangerouslySetInnerHTML={
          Object {
            "__html": "Hello",
          }
        }
      />
      <styled.small
        dangerouslySetInnerHTML={
          Object {
            "__html": "April 13, 2016",
          }
        }
      />
    </header>
    <main
      className="main"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Hello world</p>",
        }
      }
    />
  </styled.section>
</styled.div>
`;
